### PR TITLE
refactor(web): OSK spacebar-label updates now managed by layer object 🪠

### PR DESCRIPTION
**Pull request by jahorton**
To prepare to optimize layout-reflow for Web's OSK, it will help us to rework how spacebar-text and its styling are managed.  They'll be far easier to "batch" with other layout updates if managed by the layer containing the key, rather than set outside of the standard flow.

Also, we had a _lot_ of redundant calls to the `showLanguage()` function, which is what updated the spacebar.  We only need one.

This PR has been spun off of an older state of #11140.  Certain design decisions may make more sense with the whole picture in mind, though I have tried to recontextualize them to make better sense in isolation for this PR.

## User Testing

TEST_ANDROID_SPACEBAR:  Using Keyman for Android, verify that the spacebar is always appropriately captioned (based on user settings in the menu) for the active keyboard, with the text properly sized.

- If not already installed, download and test using `balochi_phonetic` - it has very high font scaling that will help stress-test part of spacebar-display management.
    - If the text overflows the spacebar at any point, overrunning its boundaries, FAIL this test.
- Test with a few other keyboards as well.
- Be sure to swap the device orientation back and forth a few times - from portrait to landscape and back again.
- Be sure to swap layers and verify that the spacebar for each is always labeled appropriately.

### CODE REVIEW

**Comment by bharanidharanj** [comment]
# Test Results

* **TEST_ANDROID_SPACEBAR (FAILED):** Tested with the attached PR build (Keyman 17.0.0303-beta-test-11175 on different Android versions (Android 5.0, 7.1, 9.0 and 12.0, 13, 14) and here is my observation: 1. Installed balochi_phonetic keyboard. 2. Installed Khmer Angkor and SIL (IPA) keyboard. 3. Switched to  Balochi_Phonetic keyboard. Verified that the Spacebar captioned appropriately for the active keyboard. 4. Changed the orientation from portrait to landscape. 5. I observed that the label name is exceeding its boundaries in Android 5, 9, 7.1, and 9 OS versions. However, it displays correctly in Android 12, 13, and 14 OS versions.

..Android 5

   

..Android 7.1

   

..Android 9.0

   

..Android 12

**Comment by jahorton** [comment]
> # Test Results
> 
> * **TEST_ANDROID_SPACEBAR (FAILED):** Tested with the attached PR build (Keyman 17.0.0303-beta-test-11175 on different Android versions (Android 5.0, 7.1, 9.0 and 12.0, 13, 14) and here is my observation: 1. Installed balochi_phonetic keyboard. 2. Installed Khmer Angkor and SIL (IPA) keyboard. 3. Switched to  Balochi_Phonetic keyboard. Verified that the Spacebar captioned appropriately for the active keyboard. 4. Changed the orientation from portrait to landscape. 5. I observed that the label name is exceeding its boundaries in Android 5, 9, 7.1, and 9 OS versions. However, it displays correctly in Android 12, 13, and 14 OS versions.
> 
> 
> ..Android 5
> 
> 
> [...]
> 
> ..Android 12
> 

I'm going to call this a pass, but I _really_ can't blame you for reporting it this way.  TL,DR: technical limitations with older Chrome versions.

TEST_ANDROID_SPACEBAR:  PASS

----

The reason:  the technique we use to ensure that the key height stays within the spacebar's boundary isn't available to us in old versions of Chrome.  Per https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics#browser_compatibility, we need a minimum of Chrome 87 in order to scale based on key-height.  (See the [`fontBoundingBoxAscent`](https://developer.mozilla.org/en-US/docs/Web/API/TextMetrics/fontBoundingBoxAscent) entry.)

Re: Android 7.1.1, I found a link suggesting that Android 7 shipped with Chrome 58.  For Android 9.0, APKMirror offers Chrome 73.  As Android Studio emulators typically start out with the minimum possible Chrome version, this explains why the height scaling wasn't working.

So, a couple of things of note:
1. No errors or crashes resulted from trying to use Web features not available given the version of Chrome in place.
2. The key _width_ scaling still worked perfectly fine.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.305-beta](https://keyman.com/downloads/releases/beta/17.0.305)

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 1]
```diff
@@ -76,6 +81,21 @@ export default class OSKLayer {
...
+
+    if(this.spaceBarKey) {
+      const spacebarLabel = this.spaceBarKey.label;
+      let tButton = this.spaceBarKey.btn;
+
+      if (typeof (tButton.className) == 'undefined' || tButton.className == '') {
+        tButton.className = 'kmw-spacebar';
+      } else if (tButton.className.indexOf('kmw-spacebar') == -1) {
+        tButton.className += ' kmw-spacebar';
+      }
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:93]:
```suggestion
      tButton.classList.add('kmw-spacebar');
```

https://developer.mozilla.org/en-US/docs/Web/API/Element/classList

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 2]
```diff
@@ -76,6 +81,21 @@ export default class OSKLayer {
...
+
+      if (spacebarLabel.className != 'kmw-spacebar-caption') {
+        spacebarLabel.className = 'kmw-spacebar-caption';
+      }
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:97]:
```suggestion
      spacebarLabel.className = 'kmw-spacebar-caption';
```

No need to guard this ... I am sure it'll already be optimized.

Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:97]: Or how about this ... so much shorter:

```suggestion
      this.spaceBarKey.btn.classList.add('kmw-spacebar');
      this.spaceBarKey.label.className = 'kmw-spacebar-caption';
```

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 3]
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
     return null;
   }
 
-  public refreshLayout(vkbd: VisualKeyboard, layerHeight: number) {
+  /**
+   * Indicate the current language and keyboard on the space bar
+   **/
+  showLanguage(displayName: string) {
+    if(!this.spaceBarKey) {
+      return () => {};
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:124]: What's this about? Nothing else returns anything?

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 4]
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
...
+      this.spaceBarKey.spec.text = displayName;
+
+      // It sounds redundant, but this dramatically cuts down on browser DOM processing;
+      // but sometimes innerText is reported empty when it actually isn't, so set it
+      // anyway in that case (Safari, iOS 14.4)
+      if (spacebarLabel.innerText != displayName || displayName == '') {
+        spacebarLabel.innerText = displayName;
+      }
+    }
+    catch (ex) { }
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:141]: Why are we catching everything here?

File: web/src/engine/osk/src/keyboard-layout/oskLayer.ts [hunk 5]
```diff
@@ -96,7 +116,35 @@ export default class OSKLayer {
...
+  }
+
+  public refreshLayout(vkbd: VisualKeyboard, layoutParams: LayerLayoutParams) {
+    const layerHeight = layoutParams.keyboardHeight;
```
Comment (mcdurdin) [inline web/src/engine/osk/src/keyboard-layout/oskLayer.ts:145]: Why is this here?

**Review by mcdurdin — COMMENTED** [review]
This is pretty much fine, but there are some funny leftover bits and pieces and some code that really needs simplification.
